proto/internal/uber/cadence/matching/v1/service.proto

```diff
@@ -0,0 +1,214 @@
+  // by the MatchingEngine.
+  rpc AddDecisionTask(AddDecisionTaskRequest) returns (AddDecisionTaskResponse);
+
+  // AddActivityTask is called by the history service when a decision task is scheduled, so that it can be dispatched
+  // by the MatchingEngine.
+  rpc AddActivityTask(AddActivityTaskRequest) returns (AddActivityTaskResponse);
+
+  // QueryWorkflow is called by frontend to query a workflow.
+  rpc QueryWorkflow(QueryWorkflowRequest) returns (QueryWorkflowResponse);
+
+  // RespondQueryTaskCompleted is called by frontend to respond query completed.
+  rpc RespondQueryTaskCompleted(RespondQueryTaskCompletedRequest) returns (RespondQueryTaskCompletedResponse);
+
+  // CancelOutstandingPoll is called by frontend to unblock long polls on matching for zombie pollers.
+  // Our rpc stack does not support context propagation, so when a client connection goes away frontend sees
+  // cancellation of context for that handler, but any corresponding calls (long-poll) to matching service does not
+  // see the cancellation propagated so it can unblock corresponding long-polls on its end.  This results is tasks
+  // being dispatched to zombie pollers in this situation.  This API is added so everytime frontend makes a long-poll
+  // api call to matching it passes in a pollerID and then calls this API when it detects client connection is closed
+  // to unblock long polls for this poller and prevent tasks being sent to these zombie pollers.
+  rpc CancelOutstandingPoll(CancelOutstandingPollRequest) returns (CancelOutstandingPollResponse);
+
+  // DescribeTaskList returns information about the target tasklist, right now this API returns the
+  // pollers which polled this tasklist in last few minutes.
+  rpc DescribeTaskList(DescribeTaskListRequest) returns (DescribeTaskListResponse);
+
+  // ListTaskListPartitions returns a map of partitionKey and hostAddress for a taskList
+  rpc ListTaskListPartitions(ListTaskListPartitionsRequest) returns (ListTaskListPartitionsResponse);
+}
+
+message PollForDecisionTaskRequest {
+  api.v1.PollForDecisionTaskRequest request = 1;
+  string domain_id = 2;
+  string poller_id = 3;
+  string forwarded_from = 4;
+}
+
+message PollForDecisionTaskResponse {
+  bytes task_token = 1;
+  api.v1.WorkflowExecution workflow_execution = 2;
+  api.v1.WorkflowType workflow_type = 3;
+  int64 previous_started_event_id = 4;
+  int64 started_event_id = 5;
+  int32 attempt = 6;
+  int64 next_event_id = 7;
+  int64 backlog_count_hint = 8;
+  bool sticky_execution_enabled = 9;
+  api.v1.WorkflowQuery query = 10;
+  shared.v1.TransientDecisionInfo decision_info = 11;
+  api.v1.TaskList workflow_execution_task_list = 12;
+  int32 event_store_version = 13;
+  bytes branch_token = 14;
+  google.protobuf.Timestamp scheduled_time = 15 [(gogoproto.stdtime) = true];
+  google.protobuf.Timestamp started_time = 16 [(gogoproto.stdtime) = true];
+  map<string, api.v1.WorkflowQuery> queries = 17;
```

Comment:
both "query" and "queries"?  I assume this mimics the thrift API, but that's a bit odd